### PR TITLE
[SDM] Enable monitoring server in android

### PR DESCRIPTION
**Pull request by djmixkim**
This patch enables monitoring server except
CheckCpuUsage function in android.

Signed-off-by: DongJun Kim <djmix.kim@samsung.com>

### CODE REVIEW

File: service_discovery_manager/Component/mmBASE/BaseAPI/ifaddrs.cc [hunk 1]
```diff
@@ -0,0 +1,221 @@
+/*
+ *  Copyright 2012 The WebRTC Project Authors. All rights reserved.
+ *
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
+ *  tree. An additional intellectual property rights grant can be found
+ *  in the file PATENTS.  All contributing project authors may
+ *  be found in the AUTHORS file in the root of the source tree.
+ */
```
Comment (younghajung) [inline service_discovery_manager/Component/mmBASE/BaseAPI/ifaddrs.cc]: This license is not full license, LICENSE file is not exist in the root of SDM. Is it OK?

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp [hunk 2]
```diff
@@ -32,6 +31,7 @@
 #endif
 
 #include "monitor_server.h"
+#include "ifaddrs.h"
```
Comment (younghajung) [inline service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: Can it work on Windows?

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp [hunk 3]
```diff
@@ -251,25 +263,52 @@ MonitorServer::MonitorServer()
   monitor_.StartMainLoop(nullptr);
   cpu_usages_.clear();
 
-#ifdef LINUX
-  // initialize for cpu usage
-  FILE* file = fopen("/proc/stat", "r");
-  fscanf(file, "cpu %llu %llu %llu %llu", &last_total_user,
-         &last_total_user_low, &last_total_sys, &last_total_idle);
-  fclose(file);
+#if defined(OS_ANDROID)
```
Comment (younghajung) [inline service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]:
```suggestion
#if defined(ANDROID)
```

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp [hunk 4]
```diff
@@ -251,25 +263,52 @@ MonitorServer::MonitorServer()
...
+  if ((file = fopen("/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq", "r")) != NULL) {
+    fscanf(file, "%lf", &frequency);
+    fclose(file);
+    frequency_ = (float)(frequency / 1000000);
+  } else {
+    DPRINT(COMM, DEBUG_ERROR,
+       "Could not open /sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq - errno(%d)\n", errno);
+    frequency_ = 1.0f;
+  }
+#elif defined(OS_LINUX)
```
Comment (younghajung) [inline service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]:
```suggestion
#elif defined(LINUX)
```

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp [hunk 5]
```diff
@@ -284,23 +323,51 @@ MonitorServer::MonitorServer(const CHAR* msg_name)
...
-  // initialize for cpu usage
-  FILE* file = fopen("/proc/stat", "r");
-  fscanf(file, "cpu %llu %llu %llu %llu", &last_total_user,
-         &last_total_user_low, &last_total_sys, &last_total_idle);
-  fclose(file);
 
-  // get cpu core
-  // TODO (djmix.kim) : How to get active cores?
-  cpu_cores_ = std::thread::hardware_concurrency();
+#if defined(OS_ANDROID)
```
Comment (younghajung) [inline service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]:
```suggestion
#if defined(ANDROID)
```

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp [hunk 6]
```diff
@@ -284,23 +323,51 @@ MonitorServer::MonitorServer(const CHAR* msg_name)
...
+  if ((file = fopen("/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq", "r")) != NULL) {
+    fscanf(file, "%lf", &frequency);
+    fclose(file);
+    frequency_ = (float)(frequency / 1000000);
+  } else {
+    DPRINT(COMM, DEBUG_ERROR,
+       "Could not open /sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq - errno(%d)\n", errno);
+    frequency_ = 1.0f;
+  }
+#elif defined(OS_LINUX)
```
Comment (younghajung) [inline service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]:
```suggestion
#elif defined(LINUX)
```

File: service_discovery_manager/Project/Android/.idea/codeStyles/Project.xml [hunk 7]
```diff
@@ -1,29 +0,0 @@
-<component name="ProjectCodeStyleConfiguration">
```
Comment (younghajung) [inline service_discovery_manager/Project/Android/.idea/codeStyles/Project.xml:1]: Please keep this file.

File: service_discovery_manager/Component/mmBASE/BaseAPI/ifaddrs.cc [hunk 8]
```diff
@@ -0,0 +1,221 @@
+/*
```
Comment (younghajung) [inline service_discovery_manager/Component/mmBASE/BaseAPI/ifaddrs.cc]: All source file of this project is cpp, so please rename to ifaddrs.cpp

File: service_discovery_manager/Project/Android/app/CMakeLists.txt [hunk 9]
```diff
@@ -36,6 +36,7 @@ add_library( # Sets the name of the library.
         ../../../Component/mmBASE/BaseAPI/bSocket.cpp
         ../../../Component/mmBASE/BaseAPI/bTask.cpp
         ../../../Component/mmBASE/BaseAPI/bThread.cpp
+        ../../../Component/mmBASE/BaseAPI/ifaddrs.cc
```
Comment (younghajung) [inline service_discovery_manager/Project/Android/app/CMakeLists.txt]: Please add source file to Component/SOURCE.LIST as well.

File: service_discovery_manager/Component/mmSH/server_runner.cpp [hunk 10]
```diff
@@ -95,14 +95,11 @@ int ServerRunner::Run() {
                                                   (void*)mh_discovery_server,
                                                   OnDiscoveryServerEvent);
 
-  // TODO: MonitorServer for Android should be implemented.
-#if !defined(ANDROID)
   MonitorServer* monitor_server = new MonitorServer(UUIDS_MDS);
```
Comment (Chikdolman) [inline service_discovery_manager/Component/mmSH/server_runner.cpp:98]: It seems that MonitorServer now supports Android. 
Therefore, MonitorServer should be freed in any case. 
I suggest to remove #if !defined(ANDROID) in line 144.

**Review by Chikdolman — COMMENTED** [review]
Please check the comment

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp (hunk 2 above)
Comment (younghajung) [reply service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: Tizen build has been failed with this patch, please check it.

File: service_discovery_manager/Component/mmSH/server_runner.cpp (hunk 10 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmSH/server_runner.cpp:98]: Thanks! I've removed line 144.

File: service_discovery_manager/Component/mmBASE/BaseAPI/ifaddrs.cc (hunk 1 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmBASE/BaseAPI/ifaddrs.cc]: added on latest patch-set

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp (hunk 2 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: move under !WIN32.

Comment (djmixkim) [reply service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: I will check tizen build and update result

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp (hunk 3 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: done

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp (hunk 4 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: done

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp (hunk 5 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: done

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp (hunk 6 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: done

File: service_discovery_manager/Project/Android/.idea/codeStyles/Project.xml (hunk 7 above)
Comment (djmixkim) [reply service_discovery_manager/Project/Android/.idea/codeStyles/Project.xml:1]: This is removed by android studio. For now, keep current changes.

File: service_discovery_manager/Component/mmBASE/BaseAPI/ifaddrs.cc (hunk 8 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmBASE/BaseAPI/ifaddrs.cc]: done

File: service_discovery_manager/Project/Android/app/CMakeLists.txt (hunk 9 above)
Comment (djmixkim) [reply service_discovery_manager/Project/Android/app/CMakeLists.txt]: done

File: service_discovery_manager/Component/mmDiscovery/monitor_server.cpp (hunk 2 above)
Comment (djmixkim) [reply service_discovery_manager/Component/mmDiscovery/monitor_server.cpp]: I've updated on latest patch-set.
android and tizen build works fine.

File: service_discovery_manager/Component/SOURCE.LIST [hunk 11]
```diff
@@ -1,44 +1,45 @@
...
-SOURCES = 	osal	            \
-		daemonAPI	    \
-		posixAPI	    \
-		socketAPI	    \
-		timeAPI	            \
-		uartAPI             \
-		processAPI					\
-		ioAPI
+SOURCES = 	daemonAPI		\
+			ioAPI			\
```
Comment (younghajung) [inline service_discovery_manager/Component/SOURCE.LIST]: Please keep indentation.

Comment (djmixkim) [reply service_discovery_manager/Component/SOURCE.LIST]: done

**Review by younghajung — APPROVED** [review]

**Pull request merged**